cache loaded python modules

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149910
* #149932
I am splitting caching the loading of modules from the caching the codegen since its trivial and much easier. 
Module loading is 50% of the cost, and codegen is 50%  of maybe_append choice on full graph model. which is 40% of total compile time.

<img width="434" alt="Screenshot 2025-03-24 at 4 35 12 PM" src="https://github.com/user-attachments/assets/aa851c6a-bde9-43f8-b12d-e439504ef62c" />


running mm_loop benchmark, 
before this change:
67947323682

after this change:
25845073249

2.6X faster. 

it seems that the cache was there then got dropped. I added benchmark so it wont be dropped again by mistake.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
cache loaded python modules

[ghstack-poisoned]

laithsakka committed:
Update on "cache loaded python modules"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "cache loaded python modules"


running mm_loop benchmark, 
before this change:

after this change:


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste reviewed (approved):
nice perf bump

laithsakka commented on torch/_inductor/codecache.py:
```diff
@@ -2861,6 +2866,10 @@ def load_by_key_path(
 
+        # we only cache whena ttrs is None
+        if attrs is None and path in cls.modules_no_attr:
+            return cls.modules_no_attr[path]
```
yeh good point , we probably need to add everything to PyCodeCache.modules also

laithsakka replied:
I am not sure when it is called with attrs, 
but this is called without attrs in this case. 

When we do autotuning of mm, we generate python code by expanding some templates then we write it to a file and load it. that loading calls this function without attrs

laithsakka replied:
I noticed that PyCodeCache.load(..) is never called with attr and linemap.

laithsakka committed:
Update on "cache loaded python modules"

I am splitting caching the loading of modules from the caching the codegen since its trivial and much easier. 
Module loading is 50% of the cost, and codegen is 50%  of maybe_append choice on full graph model. which is 40% of total compile time.

<img width="434" alt="Screenshot 2025-03-24 at 4 35 12 PM" src="https://github.com/user-attachments/assets/aa851c6a-bde9-43f8-b12d-e439504ef62c" />


running mm_loop benchmark, 
before this change:
67947323682

after this change:
25845073249

2.6X faster. 

it seems that the cache was there then got dropped. I added benchmark so it wont be dropped again by mistake.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "cache loaded python modules"

I am splitting caching the loading of modules from the caching the codegen since its trivial and much easier. 
Module loading is 50% of the cost, and codegen is 50%  of maybe_append choice on full graph model. which is 40% of total compile time.

<img width="434" alt="Screenshot 2025-03-24 at 4 35 12 PM" src="https://github.com/user-attachments/assets/aa851c6a-bde9-43f8-b12d-e439504ef62c" />


running mm_loop benchmark, 
before this change:
67947323682

after this change:
25845073249

2.6X faster. 

it seems that the cache was there then got dropped. I added benchmark so it wont be dropped again by mistake.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -2889,7 +2896,14 @@ def cache_clear(cls, purge: bool = False) -> None:
                     os.remove(mod.__file__)
                 except FileNotFoundError:
                     pass
+            for mod in cls.modules_no_attr.values():
```
I guess we don't need this part now that you've changed to always append to cls.modules?

laithsakka committed:
Update on "cache loaded python modules"

I am splitting caching the loading of modules from the caching the codegen since its trivial and much easier. 
Module loading is 50% of the cost, and codegen is 50%  of maybe_append choice on full graph model. which is 40% of total compile time.

<img width="434" alt="Screenshot 2025-03-24 at 4 35 12 PM" src="https://github.com/user-attachments/assets/aa851c6a-bde9-43f8-b12d-e439504ef62c" />


running mm_loop benchmark, 
before this change:
67947323682

after this change:
25845073249

2.6X faster. 

it seems that the cache was there then got dropped. I added benchmark so it wont be dropped again by mistake.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchbot merge
